From 1c93be57b9fd16269bb1f99cca9bfa94d494a6bc Mon Sep 17 00:00:00 2001
From: Aleksa Sarai <asarai@suse.de>
Date: Sat, 19 Nov 2016 20:25:04 +1100
Subject: [PATCH] entries: prepend all hierarchies with keyword metadata

In order to ensure that CollectUsedKeywords works on manifests as
expected (it returns the keywords the manifest was generated with),
prepend every manifest with a dummy /set and /unset entry which just
tags the manifest as "this is a keyword used".

Signed-off-by: Aleksa Sarai <asarai@suse.de>
---
 walk.go | 22 ++++++++++++++++++++++
 1 file changed, 22 insertions(+)

diff --git a/walk.go b/walk.go
index 7f5b17144cdd..c4e7d3094fcd 100644
--- a/walk.go
+++ b/walk.go
@@ -367,10 +367,32 @@ func signatureEntries(root string) []Entry {
 // keywords requested when generating this manifest.
 func keywordEntries(keywords []Keyword) []Entry {
 	// Convert all of the keywords to zero-value keyvals.
+	kvs := []string{}
+	for _, kw := range keywords {
+		kvs = append(kvs, fmt.Sprintf("%s=", string(kw)))
+	}
+
+	// Convert all of the keywords to zero-value keyvals.
 	return []Entry{
 		{
 			Type: CommentType,
 			Raw:  fmt.Sprintf("#%16s%s", "keywords: ", strings.Join(FromKeywords(keywords), ",")),
 		},
+		{
+			Type: BlankType,
+		},
+		{
+			Type: CommentType,
+			Raw:  "# <keywords>",
+		},
+		{
+			Type:     SpecialType,
+			Name:     "/set",
+			Keywords: StringToKeyVals(kvs),
+		},
+		{
+			Type: SpecialType,
+			Name: "/unset",
+		},
 	}
 }
-- 
2.10.2

